### PR TITLE
Improve Netlify CMS integration

### LINKED ISSUE #262: Netlify CMS Improvement: Deploy preview links
## Issue description

Currently, the deploy preview link on Netlify CMS opens the **home** page of the preview and a user has to navigate to the topic they changed. This improvement will open the preview on the page the user is currently editing.

It's a bad user experience to send the user to the home page of the docs, and make them to navigate to the topic that they've updated so they can see the preview. It can be challenging, confusing, and frustrating for the user to find the topic.

## Problem location

Preview link on Netlify CMS

## Suggestions for a fix

I've discovered that today is not possible to fix the links in our project. So I've opened this ticket  [#3084](https://github.com/netlify/netlify-cms/issues/3084) on Netlify CMS project, and they will implement a solution. When they do it, we'll have to check how to implement their solution in our project.

### LINKED ISSUE #271: Update Netlify CMS version
Update CMS version to avail of the fix for forks not being updated https://github.com/netlify/netlify-cms/issues/2962

### LINKED ISSUE #272: Netlify CMS improvement: Edit on Netlify CMS links should open the exact page to edit
### Improvement
Currently, the **Edit on Netlify CMS** link opens the home page of the CMS admin and a user has to navigate to the topic they want to edit. This improvement will open the exact page the user is reading when they click the link (same UX as Github edit link).
image

### Task
Update the code for the Edit on Netlify CMS link to use the edit path for the file in the CMS as per https://www.netlifycms.org/docs/open-authoring#linking-to-specific-entries-in-the-cms

### Notes
An attempt was made to fix this in PR #203 but it was closed due to another issue with the CMS.

### PR DESCRIPTION
**Pull request by erquhart**
Hello there!

I noticed you've been using our Open Authoring beta for some time, and thought I'd use your site to put the new feature through the paces and see a.) how well it holds up with multi-level content nesting, and b.) how much effort proper customization takes.

This PR includes:
- Direct "Edit on Netlify CMS" links to the current entry
- Customize preview - see actual site styling in the preview pane while editing
- Move configuration to js and reuse default collection configuration
- Add editor component for Docsy `Alert` shortcode for visual editing and preview of alerts

Hope it helps, feel free to remove anything you don't like!

fixes #262 fixes #271 fixes #272

### CODE REVIEW

File: netlify.toml [hunk 1]
```diff
@@ -1,2 +1,6 @@
 [build.environment]
   HUGO_VERSION = "0.55.6"
+
+[dev]
+  targetPort = 1313
+  port = 1314
```
Comment (erquhart) [inline netlify.toml]: Assign a port when using Netlify Dev for local development.

**Comment by erquhart** [comment]
Looks like there's a hash in the production css filename, stopping it from being loaded in the preview pane. Need to look into how to get that filename into the CMS.

**Comment by alexearnshaw** [comment]
@erquhart Wow, thanks so much for taking the time to do this! Having the "Edit on Netlify CMS" link directly to the topic in the CMS is something that's been on our wish list for some time, but we didn't really know how best to implement it. Some other great improvements here too :) 

I've had a brief look at the changes and don't see any issues, but I'd like to take a more thorough look and do some testing, which I hope to do in the next few days.

**Comment by alexearnshaw** [comment]
> Looks like there's a hash in the production css filename, stopping it from being loaded in the preview pane. Need to look into how to get that filename into the CMS.

I'm not sure I understand the issue, but I think it means we won't see the site styling in the preview pane unless you get this working?

**Comment by erquhart** [comment]
Right, I'm trying to inject the css filename as `<filename>.css`, but the actual file is `<filename><really-long-hash>.css`. Still need to circle back, will do soon 👍

**Comment by erquhart** [comment]
@alexearnshaw this is good to go 👍

**Comment by alexearnshaw** [comment]
@erquhart Have you any tips on the easiest way to verify all your changes before I merge them? We don't have any automated testing or anything in place and we currently don't have any developers on this project, just some tech writers with limited development experience ;)

I've tried to test using the Netlify deploy preview (modifying the CMS URLs when necessary) and I can see that the CMS preview pane is now using our site styling, and that the Alert component editing and preview is there, which looks great! 

But I am seeing some issues with the other changes, e.g. the Edit on Netlify CMS button is just opening the CMS home page instead of the entry even though I can see by the URL it's linking to that it should be opening the exact entry. Also all images are now showing as broken in the preview pane in the CMS:

image

I think this is just not going to work properly until I merge it as I've had similar issues testing CMS changes on deploy previews before. I've also tried pulling down your fork/branch locally and running on localhost but I see the same issues. I have a test fork that I can merge your changes into, but just wanted to check if there was an easier way?

**Comment by alexearnshaw** [comment]
Update: I've merged your branch into my test fork of the repo (https://github.com/alexearnshaw/axway-open-docs) and am still  seeing the issues with the Edit on Netlify CMS links and broken images in the CMS preview...

**Comment by erquhart** [comment]
Weird, the images were working for me - looking into it now

**Comment by erquhart** [comment]
Looks like we have some work to do on absolute pathed images in our latest release. I'll close this for now and re-open when we're ready.

**Comment by erquhart** [comment]
Update: still working on this, ironing out CMS issues uncovered in this implementation.

**Comment by alexearnshaw** [comment]
Thanks for the update @erquhart

**Comment by alexearnshaw** [comment]
@erquhart Any approx ETA on when the image issues will be addressed. I tried to update to latest CMS version yesterday as we really want to avail of the bug fix for forks (https://github.com/netlify/netlify-cms/issues/2962) as many of our users are having issues caused by this. But I encountered a couple of different image issues (see #302 for details) so thought I better hold off...

**Comment by erquhart** [comment]
@alexearnshaw missed this comment somehow - reopening this in a bit.

**Comment by erquhart** [comment]
Where this PR stands:

**Images**
- Images are mostly displaying in the CMS
- Collections can only show images from one source - a folder has been configured for each based on current usage
- For collections that have images from multiple directories, the images that are not in the configured folder do not show in the CMS, but the site itself is not impacted
- The CMS will need to be able to handle file hierarchies of images in the media library to properly address this

**Edit links**
- I'm releasing a fix right now in the CMS that will get edit links working again
- Still need to see if _index pages redirect properly

**Preview paths**
- Added preview path configuration, so the "View Live" and "View Preview" links in the top right will go straight to the page being edited

**Comment by erquhart** [comment]
Looking in to an issue with edit links that's happening in production builds only.

**Comment by alexearnshaw** [comment]
Thanks @erquhart for the comprehensive update, this is going to make a huge improvement to usability when it's complete.

**Comment by erquhart** [comment]
@alexearnshaw I think this is good to go. There are two outstanding issues:
- https://github.com/netlify/netlify-cms/issues/3241 - will likely be fixed in short order. Because your docs are pulling the CMS direct from the unpkg.com CDN, you'll automatically get the fix when released.
- https://github.com/netlify/netlify-cms/issues/3240 - this should completely solve missing images in your CMS UI by properly handling nested directories of images, but it's a bit of a lift and won't be completed immediately. It is definitely a priority, though.

My two cent: I don't think waiting on either of these fixes is a requirement for you, as this PR should bring considerable improvement to your current editing experience.

Please comment if you find any further issues with this PR. And if you find bugs or opportunities for improvement, please [open an issue](https://github.com/netlify/netlify-cms/issues/new)!

**Comment by alexearnshaw** [comment]
@erquhart Fantastic, thank you. I plan to spend some time tomorrow reviewing and testing this but I agree that we don't need to wait for those 2 fixes to go ahead and merge this PR.

And thanks a lot for noticing the missing collections in the CMS config, this was a complete oversight on our part that we might not have noticed.

**Comment by alexearnshaw** [comment]
fixes #262 #271 #272

**Pull request merged**